### PR TITLE
feat: 損失・学習ループ・公式 main（losses / trainer / main）

### PR DESCRIPTION
**Pull request by koki3070**
## 概要

PR5（距離バックエンド・持続ホモロジー）の上に、**損失・学習ループ・公式エントリポイント**を追加する PR です（**分割 PR の 6 本目**）。

- **`tda_ml/losses.py`:** 分類 BCE、楕円サイズ／異方性正則化、位相 Wasserstein 損失（`distance_backend` 対応）。
- **`tda_ml/trainer.py`:** 学習・検証ループ、AMP、可視化フック、clean PD 計算。
- **`tda_ml/main.py`:** `NoisyMNISTDataset` + `Trainer` による公式学習入口（`python -m tda_ml.main --config dev` 等）。
- **`tda_ml/visualization.py`:** エポックごとの GT／予測／楕円プロット。
- **テスト:** `tests/test_trainer.py`, `tests/test_losses_topo_distance.py`。

**Depends on #56**（PR5）。未マージ時は PR1〜5 の変更も差分に含まれる場合があります。

## 楕円パラメータ

- **`points`:** `(N, 2)` 各点の座標。
- **`params`:** `(N, 3)` 各点の `[a, b, theta]`（中心は `points`）。

## 位相損失とバックエンド

- **Mahalanobis:** `TopologicalLoss` は `sigmoid(logits)` を outlier 確率として距離に織り込む。
- **ellphi:** 確率重みは未実装（`distance_backend` と同様）。公平比較の注意は PR7 の README 参照。

## この PR に含めないもの

- `experiments/run_backend_multiseed.py`, `REPRODUCIBILITY.md`（**PR7**）
- CI の `pytest` / 再現スモーク（引き続き **ruff のみ**）

## テスト計画

- [ ] CI: `Ruff` 成功
- [ ] （任意）`uv run pytest tests/test_trainer.py tests/test_losses_topo_distance.py`
- [ ] （任意）`uv run python -m tda_ml.main --config dev`（3 epoch スモーク）

### CODE REVIEW

**Comment by t-uda** [comment]
Codex review item 3（NaN skip 後に全 batch skip だと metrics が空入力になり得る件）については、確かに制御フロー上の edge case ではありますが、私は **merge blocker ほどではない** と見ています。

現状の `train_epoch()` は NaN loss の batch を skip し、epoch 末で train metrics を計算します。したがって、理屈の上では「その epoch の全 batch が NaN で skip された場合」に `all_train_labels` / `all_train_preds` が空のまま sklearn metrics に渡る可能性があります。

ただし、これは通常の設定で頻繁に起きる failure mode ではなく、数値的に完全に壊れた config / backend の検出に近い状況です。その意味では、研究実験の安全性としては優先度 medium 以下でよいと思います。

軽く直すなら、epoch 末の metrics 計算前に

```python
if steps_completed == 0 or not all_train_labels:
    raise RuntimeError(
        f"All training batches were skipped at epoch={epoch}; loss was NaN for every batch."
    )
```

程度の明示的な controlled failure を入れるのが十分です。NaN metrics を無理に返して training を継続するより、ここまで壊れた場合は明示的に落とす方が分かりやすいと思います。

**Comment by koki3070** [comment]
epoch 末の train metrics 計算前に、全 batch が NaN skip された場合は `RuntimeError` で明示的に落とすようにしました。通常運用では起きにくい edge case ですが、空入力で metrics を計算するより分かりやすい挙動にしています。

File: tda_ml/losses.py [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+            
+        axes = params[..., 0:2]
+        major_axis = axes.max(dim=-1)[0]
+        minor_axis = axes.min(dim=-1)[0]
+        
+        aspect_ratios = major_axis / (minor_axis + 1e-6)
+        
+        if self.mode == 'barrier':
+            barrier_term = F.relu(aspect_ratios - self.barrier_threshold).pow(2).mean()
+            loss = 10.0 * barrier_term
```
Comment (t-uda) [inline tda_ml/losses.py:116]: #60

File: tda_ml/trainer.py [hunk 2]
```diff
@@ -0,0 +1,319 @@
...
+            import random
+            self.fixed_indices = random.sample(range(len(data_loader.dataset)), 3)
+
+        for i, (data, labels, clean_pc) in enumerate(pbar):
+            data = data.to(self.device, non_blocking=True)
+            labels = labels.to(self.device, non_blocking=True)
+            clean_pc = clean_pc.to(self.device, non_blocking=True)
+
+            if self.config.get('training', {}).get('rotation_augmentation', False):
+                theta = torch.rand(1, device=self.device) * 2 * 3.141592653589793
```
Comment (t-uda) [inline tda_ml/trainer.py:145]: #60

**Review by t-uda — APPROVED** [review]
LGTM!

**Comment by t-uda** [comment]
スカッシュコミット da38fc4 を main に直接 push して適用済み（PR#56 同様、squash merge 累積によるコンフリクト回避のためローカルでマージ）。
